File: packages/types/typescript/translateEnums.ts
import * as data from "./generated/data";

export class EnumTranslator {
  private static instance: EnumTranslator;

  private constructor() {}

  public static getInstance(): EnumTranslator {
    if (!EnumTranslator.instance) {
      EnumTranslator.instance = new EnumTranslator();
    }
    return EnumTranslator.instance;
  }

  // --- RacePhase ---
  public racePhaseEnumToString(value: data.RacePhase): string {
    const str = data.racePhaseToJSON(value);
    return str === "UNRECOGNIZED" ? "phase_unknown" : str.toLowerCase();
  }

  public racePhaseStringToEnum(str: string): data.RacePhase {
    return data.racePhaseFromJSON(str.toUpperCase());
  }

  // --- RaceCourse ---
  public raceCourseEnumToString(value: data.RaceCourse): string {
    const str = data.raceCourseToJSON(value);
    return str === "UNRECOGNIZED" ? "track_unknown" : str.toLowerCase();
  }

  public raceCourseStringToEnum(str: string): data.RaceCourse {
    return data.raceCourseFromJSON(str.toUpperCase());
  }

  // --- LapNumber ---
  public lapNumberEnumToString(value: data.LapNumber): string {
    const str = data.lapNumberToJSON(value);
    return str === "UNRECOGNIZED" ? "lap_unknown" : str.toLowerCase();
  }

  public lapNumberStringToEnum(str: string): data.LapNumber {
    return data.lapNumberFromJSON(str.toUpperCase());
  }

  // --- Placement ---
  public placementEnumToString(value: data.Placement): string {
    const str = data.placementToJSON(value);
    return str === "UNRECOGNIZED" ? "place_unknown" : str.toLowerCase();
  }

  public placementStringToEnum(str: string): data.Placement {
    return data.placementFromJSON(str.toUpperCase());
  }

  // --- Item ---
  public itemEnumToString(value: data.Item): string {
    const str = data.itemToJSON(value);
    return str === "UNRECOGNIZED" ? "item_unknown" : str.toLowerCase();
  }

  public itemStringToEnum(str: string): data.Item {
    return data.itemFromJSON(str.toUpperCase());
  }

  // --- NumCoins ---
  public numCoinsEnumToString(value: data.NumCoins): string {
    const str = data.numCoinsToJSON(value);
    return str === "UNRECOGNIZED" ? "coin_unknown" : str.toLowerCase();
  }

  public numCoinsStringToEnum(str: string): data.NumCoins {
    return data.numCoinsFromJSON(str.toUpperCase());
  }

  // --- RaceCourseType ---
  public raceCourseTypeEnumToString(value: data.RaceCourseInfo_RaceCourseType): string {
    const str = data.raceCourseInfo_RaceCourseTypeToJSON(value);
    return str === "UNRECOGNIZED" ? "course_type_unknown" : str.toLowerCase();
  }

  public raceCourseTypeStringToEnum(str: string): data.RaceCourse_RaceCourseType {
    return data.raceCourseInfo_RaceCourseTypeFromJSON(str.toUpperCase());
  }
}

export const enumTranslator = EnumTranslator.getInstance();

File: packages/types/python/translate_enums.py
import os
import sys
try:
    import data_pb2
except:
    import utils.packages.types.python.generated.data_pb2 as data_pb2
import os

# Add generated directory to path
sys.path.append(os.path.join(os.path.dirname(__file__), "generated"))


class EnumTranslator:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._init_mappings()
        return cls._instance

    def _init_mappings(self):
        # We will build bidirectional maps for each enum defined in the proto
        self.enums = {
            "RacePhase": data_pb2.RacePhase,
            "RaceCourse": data_pb2.RaceCourse,
            "LapNumber": data_pb2.LapNumber,
            "Placement": data_pb2.Placement,
            "Item": data_pb2.Item,
            "NumCoins": data_pb2.NumCoins,
            "RaceCourseType": data_pb2.RaceCourseInfo.RaceCourseType,
        }

        self.to_str = {}
        self.to_enum = {}

        for name, enum_type in self.enums.items():
            self.to_str[name] = {val: key.lower() for key, val in enum_type.items()}
            self.to_enum[name] = {key.lower(): val for key, val in enum_type.items()}

    # --- RacePhase ---
    def racePhaseEnumToString(self, value):
        return self.to_str["RacePhase"].get(value, "phase_unknown")

    def racePhaseStringToEnum(self, string):
        return self.to_enum["RacePhase"].get(string.lower(), data_pb2.PHASE_UNKNOWN)

    # --- RaceCourse ---
    def raceCourseEnumToString(self, value):
        return self.to_str["RaceCourse"].get(value, "track_unknown")

    def raceCourseStringToEnum(self, string):
        return self.to_enum["RaceCourse"].get(string.lower(), data_pb2.TRACK_UNKNOWN)

    # --- LapNumber ---
    def lapNumberEnumToString(self, value):
        return self.to_str["LapNumber"].get(value, "lap_unknown")

    def lapNumberStringToEnum(self, string):
        return self.to_enum["LapNumber"].get(string.lower(), data_pb2.LAP_UNKNOWN)

    # --- Placement ---
    def placementEnumToString(self, value):
        return self.to_str["Placement"].get(value, "place_unknown")

    def placementStringToEnum(self, string):
        return self.to_enum["Placement"].get(string.lower(), data_pb2.PLACE_UNKNOWN)

    # --- Item ---
    def itemEnumToString(self, value):
        return self.to_str["Item"].get(value, "item_unknown")

    def itemStringToEnum(self, string):
        return self.to_enum["Item"].get(string.lower(), data_pb2.ITEM_UNKNOWN)

    # --- NumCoins ---
    def numCoinsEnumToString(self, value):
        return self.to_str["NumCoins"].get(value, "coin_unknown")

    def numCoinsStringToEnum(self, string):
        return self.to_enum["NumCoins"].get(string.lower(), data_pb2.COIN_UNKNOWN)

    # --- RaceCourseType ---
    def raceCourseTypeEnumToString(self, value):
        return self.to_str["RaceCourseType"].get(value, "course_type_unknown")

    def raceCourseTypeStringToEnum(self, string):
        return self.to_enum["RaceCourseType"].get(
            string.lower(), data_pb2.RaceCourseInfo.COURSE_TYPE_UNKNOWN
        )


# Singleton instance
enumTranslator = EnumTranslator()
